ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Stream.scala

```diff
@@ -339,7 +339,9 @@ private[h2] class H2Stream[F[_]: Concurrent](
             if (needsWindowUpdate && !isClosed && sizeReadOk) {
               enqueue.offer(Chunk.singleton(H2Frame.WindowUpdate(id, windowSize - newSize)))
             } else Applicative[F].unit
-          _ <- if (data.endStream) s.trailWith(List.empty).void else Applicative[F].unit
+          _ <-
+            if (data.endStream) s.readBuffer.close *> s.trailWith(List.empty).void
```

Comment:
Probably it's kind of bikeshedding, but is it intentional that `H2Stream#receiveData` doesn't close `readBuffer` on cancelation?

Reply:
It's a good question, cancelation is handled in various places by sending an exception to the `readBuffer`.
https://github.com/http4s/http4s/blob/a1687d9dd529aa443ff23e5ea4fcce20ffac7893/ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Stream.scala#L440-L448

Reply:
Yeah, but does it work for `receiveData` as well?

Reply:
`receiveData` is called only from the `readLoop`. The `readLoop` runs forever when you create a server or a client: the only way to cancel it is to close the server/client resource. But to close the server/client resource you must have first closed all on-going requests.

https://github.com/http4s/http4s/blob/fe803615f365555693d79cfc6ce9079fa6cf63f6/ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Connection.scala#L204

Reply:
Thanks for the explanation. I can't beat my feeling suspicious about this. But this kind of leak perhaps would be noticeable.